### ISSUE TITLE
Error: JWT must have 3 parts

### ISSUE THREAD

**Issue by smiltenburg** [NakedObjectsGroup/NakedObjectsFramework#97]
Hi,

I'm trying to get Auth0 configured correctly for my NOF9 migration, but I guess I'm still missing something.
I've created an Auth0 account, put client_id / client secret in the configuration and in the the dialog pops up. In the Auth0 log section on their website I am seeing a succesfull login.

But when the login is complete I see a completely dark screen (no menus) with the following javascript errors: 

To be sure I also tried this with a "clean" Template project but here I have the same results.

main.07cb7933fa751f1f9597.bundle.js:1 **JWT must have 3 parts**

Error: JWT must have 3 parts
    at t.W4CS.t.decodeToken (http://localhost:5001/vendor.d2d67c65b8aaa0d92f6d.bundle.js:406:82139)
    at t.W4CS.t.getTokenExpirationDate (http://localhost:5001/vendor.d2d67c65b8aaa0d92f6d.bundle.js:406:82341)
    at t.W4CS.t.isTokenExpired (http://localhost:5001/vendor.d2d67c65b8aaa0d92f6d.bundle.js:406:82496)
    at r (http://localhost:5001/vendor.d2d67c65b8aaa0d92f6d.bundle.js:406:76775)
    at e.BFvY.e.authenticated (http://localhost:5001/main.07cb7933fa751f1f9597.bundle.js:1:81878)
    at Object.eval [as updateDirectives] (ng:///t/t.ngfactory.js:41:30)
    at Object.updateDirectives (http://localhost:5001/vendor.d2d67c65b8aaa0d92f6d.bundle.js:127:8855)
    at ur (http://localhost:5001/vendor.d2d67c65b8aaa0d92f6d.bundle.js:127:4471)
    at br (http://localhost:5001/vendor.d2d67c65b8aaa0d92f6d.bundle.js:127:7308)
    at Mr (http://localhost:5001/vendor.d2d67c65b8aaa0d92f6d.bundle.js:127:6706)

A google query for "JWT must have 3 parts" leads me to this recent github discussion: https://github.com/auth0/angular2-jwt/issues/335

Where it is suggested to rename the localstorage "id_token"  to "token". but after I changed this I still got the same error.

What seemes strange to me is that the "id_token" / "token"  field contains the value "undefined".

Any help in resolving this would be appreciated

**Comment by richardpawson** [comment]
Not much consolation, but I just tried it and got the same error as you.  Will investigate further and get back to you.

**Comment by richardpawson** [comment]
Sorry for the delay  -  I haven't forgotten.  I am no longer seeing this issue, and am wondering if it was some temporary glitch at Auth0, though I hope not!  But in trying to reproduce it I was seeing another issue.  The NOF Developer Manual implies that you can switch on Auth0 authentication on the client, without necessarily switching it on on the server -  though this would not be a very secure system!  But when I tried this while trying to reproduce the 'JWT must have 3 parts' error, I found that this was not the case  -  enforcing authentication in the client _without_ also enforcing it on the server, seems to lead to a 500 server error. It was working fine, though, when I had it switched on both sides (which is obviously the correct thing to do for deployment, anyway).  We will investigate further and update the Manual if appropriate.

**Comment by smiltenburg** [comment]
Hi Richard, 

Thx for investigating. However I cannot confirm the problem has also vanished for me :(

My steps to reproduce:
- start with a clear template
- removed  the // from [Authorize] on the restfulobjectscontroller
- set authenticate to true in config.json
- copied the authClientId, authDomain and applicationName from the Auth0 clients setting page to the config.json file
- added the following to allowed callback urls in the Auth0 setting page http://localhost:5001/, https://localhost:5001/
- run the application

When opening the webpage in crome INCOGNITO MODE (to clear session/cookie data from previous login attempts) I see a welcome page. I then click login. The oAuth asks me to login with my google account, which i do. Then I get returned to the welcome screen which asks me to login (and then I see the javascript errors on the page). When I press the "login" button a second time I get forwarded to a black page. (with the same javascript errors)

**Comment by richardpawson** [comment]
Well you appear to be doing everything correctly.  I went through that this morning and it worked OK for me  -  though I admit that when I tried it last week I did see the same error as you had.  

Searching online I found [this  ](https://github.com/auth0-samples/auth0-angularjs2-systemjs-sample/issues/39) suggesting that the issue is something to do with stale tokens  -  though I note you used Chromes icognito mode. But more than one person reports that the issue is intermittent.

I will ask Stef to take a look at this one and see if there is anything from that (and similar postings elsewhere) suggests anything that we should be doing differently.

**Comment by scascarini** [comment]
I created a new client on Auth0 and it appears that some of the defaults have changed from when we created our NOG test Auth0 client.  The template solution does not work with the new defaults although I don't see exactly the same error. 

I suggest you try updating your Auth0 client as follows (these are the differences on a new Auth0 client from the NOG test client) 

"Use Auth0 instead of the IdP to do Single Sign On" = false

Advanced Settings, 

OAuth,  

"JsonWebToken Signature Algorithm" = HS256
"OIDC Conformant" = false

Grant Types 

enable everything

We already have a ticket to update to use RS256, I think we will also need a ticket to update so that we run out of the box against the Auth0 defaults. I'll need to do some further investigation on that.
